cpp/src/arrow/python/udf.cc

```diff
@@ -0,0 +1,234 @@
+  if (!expected.Equals(actual)) {
+    return Status::TypeError("Expected output type, ", expected.name(),
+                             ", but function returned type ", actual.name());
+  }
+  return Status::OK();
+}
+
+// struct PythonUdf {
+//   std::shared_ptr<OwnedRefNoGIL> function;
+//   compute::OutputType output_type;
+
+//   // function needs to be destroyed at process exit
+//   // and Python may no longer be initialized.
+//   ~PythonUdf() {
+//     if (_Py_IsFinalizing()) {
+//       function->detach();
+//     }
+//   }
+
+//   Status operator()(compute::KernelContext* ctx, const compute::ExecBatch& batch,
+//                     Datum* out) {
+//     return SafeCallIntoPython([=]() -> Status { return Execute(ctx, batch, out); });
+//   }
+
+//   Status Execute(compute::KernelContext* ctx, const compute::ExecBatch& batch,
+//                  Datum* out) {
+//     const auto num_args = batch.values.size();
+//     PyObject* arg_tuple = PyTuple_New(num_args);
+//     for (size_t arg_id = 0; arg_id < num_args; arg_id++) {
+//       switch (batch[arg_id].kind()) {
+//         case Datum::SCALAR: {
+//           auto c_data = batch[arg_id].scalar();
+//           PyObject* data = wrap_scalar(c_data);
+//           PyTuple_SetItem(arg_tuple, arg_id, data);
+//           break;
+//         }
+//         case Datum::ARRAY: {
+//           auto c_data = batch[arg_id].make_array();
+//           PyObject* data = wrap_array(c_data);
+//           PyTuple_SetItem(arg_tuple, arg_id, data);
+//           break;
+//         }
+//         default:
+//           auto datum = batch[arg_id];
+//           return Status::NotImplemented(
+//               "User-defined-functions are not supported for the datum kind ",
+//               datum.ToString(batch[arg_id].kind()));
+//       }
+//     }
+//     PyObject* result;
+//     result = PyObject_CallObject(function->obj(), arg_tuple);
+//     RETURN_NOT_OK(CheckPyError());
+//     if (result == Py_None) {
+//       return Status::Invalid("Output is None, but expected an array");
+//     }
+//     // unwrapping the output for expected output type
+//     if (is_scalar(result)) {
+//       ARROW_ASSIGN_OR_RAISE(auto val, unwrap_scalar(result));
+//       RETURN_NOT_OK(CheckOutputType(*output_type.type(), *val->type));
+//       *out = Datum(val);
+//       return Status::OK();
+//     } else if (is_array(result)) {
+//       ARROW_ASSIGN_OR_RAISE(auto val, unwrap_array(result));
+//       RETURN_NOT_OK(CheckOutputType(*output_type.type(), *val->type()));
+//       *out = Datum(val);
+//       return Status::OK();
+//     } else {
+//       return Status::TypeError("Unexpected output type: ", Py_TYPE(result)->tp_name,
+//                                " (expected Scalar or Array)");
+//     }
+//     return Status::OK();
+//   }
+// };
+
+struct PythonUdf {
+  ScalarUdfWrapperCallback cb;
+  std::shared_ptr<OwnedRefNoGIL> function;
```

Comment:
Why does this have to be a shared_ptr?

Reply:
`OwnedRefNoGIL` is move-only and the function registry stores things in `std::function` which requires copies

Reply:
Ah, thanks.